pkg/common/aws/ec2.go

```diff
@@ -79,3 +85,49 @@ func (CcsAwsSession *ccsAwsSession) ReleaseElasticIPs(dryrun bool, sendSummary b
 
 	return nil
 }
+
+// TerminateEC2Instances finds EC2 instances with tag Name "osde2e*",
+// older than given duration, then terminates these EC2 instances.
+// Ignores EC2 instances with tag Name "osde2e-proxy*".
+func (CcsAwsSession *ccsAwsSession) TerminateEC2Instances(olderthan time.Duration, dryrun bool, sendSummary bool,
+	terminatedCounter *int, failedCounter *int, errorBuilder *strings.Builder,
+) error {
+	err := CcsAwsSession.GetAWSSessions()
+	if err != nil {
+		return err
+	}
+	result, err := CcsAwsSession.ec2.DescribeInstances(&ec2.DescribeInstancesInput{})
+	if err != nil {
+		return err
+	}
+	for _, reservation := range result.Reservations {
+		for _, instance := range reservation.Instances {
+			if time.Since(*instance.LaunchTime) > olderthan {
+				for _, tag := range instance.Tags {
+					if *tag.Key == "Name" && !strings.Contains(*tag.Value, osde2eproxysubstr) && strings.Contains(*tag.Value, osde2esubstr) {
+						fmt.Printf("Instance will be terminated: %s\n", instance)
+						if !dryrun {
+							input := &ec2.TerminateInstancesInput{
+								InstanceIds: []*string{instance.InstanceId},
+							}
+							_, err := CcsAwsSession.ec2.TerminateInstances(input)
+							if err != nil {
+								*failedCounter++
+								errorMsg := fmt.Sprintf("Instance %s failed to terminate: %s\n", *instance.InstanceId, err.Error())
+								fmt.Println(errorMsg)
+								if sendSummary && errorBuilder.Len() < config.SlackMessageLength {
+									errorBuilder.WriteString(errorMsg)
+								}
+								continue
+							}
+							*terminatedCounter++
+							fmt.Println("Terminated")
+						}
+						break
+					}
+				}
+			}
+		}
```

Comment:
Parts of this can be simplified by flipping your logic around. Rather than `if instance is older than N let's delete it`, you can think `if instance is not older than N let's skip it`. Same for the tag key/value containing the substring. After that, I recommend inlining any variables that you can and only printing once (unless we fail).

```suggestion
    for _, reservation := range result.Reservations {
        for _, instance := range reservation.Instances {
            if time.Since(*instance.LaunchTime) < olderthan {
                continue
            }
            for _, tag := range instance.Tags {
                if *tag.Key == "Name" && strings.Contains(*tag.Value, osde2eproxysubstr) {
                    continue
                }
                fmt.Printf("Terminating instance: %s\n", instance)
                if !dryrun {
                    _, err = CcsAwsSession.ec2.TerminateInstances(&ec2.TerminateInstancesInput{InstanceIds: aws.StringSlice([]string{*instance.InstanceId})})
                    if err != nil {
                        *failedCounter++
                        errorMsg := fmt.Sprintf("Instance %s failed to terminate: %s\n", *instance.InstanceId, err.Error())
                        fmt.Println(errorMsg)
                        if sendSummary && errorBuilder.Len() < config.SlackMessageLength {
                            errorBuilder.WriteString(errorMsg)
                        }
                        break
                    }
                    *terminatedCounter++
                }
            }
        }
```

Reply:
```
              if *tag.Key == "Name" && strings.Contains(*tag.Value, osde2eproxysubstr) {
                    continue
                }
```
The tags are a list and they seem to be in random order. If an EC2 instance does belong to osde2e-proxy but the Name tag is not the first in tag list then that EC2 instance would be deleted. This shouldn't happen I think. 
There should be an if condition(not the if not) here right?

Reply:
Let's sync tomorrow and think about rewriting this. I think we should try something like looping over to build up a list of instances to delete and provide that as one slice to the terminateinstances param